Jetpack Offer Reset: Refactor Calls to Translate Out of Constants File

robertf4 opened the pull request:
#### Changes proposed in this Pull Request

This is a followup to https://github.com/Automattic/wp-calypso/pull/45727, which was made to quickly fix an issue where usage of translated strings was not reactive. Although that solution worked, it is not ideal in the long term.

This PR re-works the Offer Reset copy that was being passed to `translate()` in the constants file so that they are instead retrieved through a function in `utils` or `translated-copy` that requires passing `translate`. Components that use this method will ideally use the `useTranslate()` hook to pass `translate`, and thus retrieve the copy in a reactive way.

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

1. Checkout the branch and start up Calypso and Jetpack Cloud.
2. In Calypso, visit the plans page. Verify that everything looks the same as it does on the production version of the plans page. Be sure to check the flows that include the details page and the upsell page.
3. Visit the plans page in another language and verify that every string on the page is translated.
4. In Jetpack Cloud, visit the pricing page. Verify that everything looks the same as it does on the production version of the pricing page.
5. Visit a localized version such as http://cloud.localhost:3001/es/pricing and verify that every string on the page is translated.

Fixes #1169247016322522-as-1194900844047779

robertf4 committed:
Refactor all translated copy out of plans-v2/constants

rcanepa commented:
In Jetpack Cloud, I'm seeing an error in the console every time I click on a CTA. It also happens when I change the product type filter. It seems to be the same error in every case.

![Kapture 2020-10-06 at 11 24 20](https://user-images.githubusercontent.com/3418513/95214869-b7975780-07c6-11eb-8cb6-0edb07fdbc0c.gif)

rcanepa commented:
The same happens to me in Calypso. The error looks like:
```
react_devtools_backend.js:2273 The above error occurred in the <UpsellComponent> component:
    in UpsellComponent (created by UpsellPage)
    in UpsellPage
    in Unknown
    in div (created by Layout)
    in div (created by Layout)
    in div (created by Layout)
    in Layout (created by ConnectFunction)
    in ConnectFunction (created by ProviderWrappedLayout)
    in MomentProvider (created by ConnectFunction)
    in ConnectFunction (created by ProviderWrappedLayout)
    in Provider (created by ProviderWrappedLayout)
    in I18nProvider (created by CalypsoI18nProvider)
    in CalypsoI18nProvider (created by ProviderWrappedLayout)
    in ProviderWrappedLayout

Consider adding an error boundary to your tree to customize error handling behavior.
Visit https://fb.me/react-error-boundaries to learn more about error boundaries.

Uncaught Unknown SelectorProductSlug: jetpack_backup_realtime
```

robertf4 commented:
Good catch @rcanepa, it seems I had missed some slugs. I've updated the PR to include the missing slugs and fixed some other things I found. I think everything should be good to go now.

a8ci18n commented on client/my-sites/plans-v2/translated-copy.tsx:
```diff
@@ -0,0 +1,203 @@
+						em: <em />,
+					},
+				} ),
+				shortName: translate( 'Backup Daily', {
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **20** times:
`translate( 'Jetpack Backup Daily' )` *ES Score: 8*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/fix/move-plansv2-translates-from-constants-take-2)

a8ci18n commented on client/my-sites/plans-v2/translated-copy.tsx:
```diff
@@ -0,0 +1,203 @@
+						em: <em />,
+					},
+				} ),
+				shortName: translate( 'Backup Real-time', {
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **20** times:
`translate( 'Jetpack Backup Real-Time' )` *ES Score: 8*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/fix/move-plansv2-translates-from-constants-take-2)

a8ci18n commented on client/my-sites/plans-v2/translated-copy.tsx:
```diff
@@ -0,0 +1,203 @@
+						'{{em}}Available options: Real-time or Daily.{{/em}}',
+					{ components: { em: <em /> } }
+				),
+				buttonLabel: translate( 'Get Jetpack Security' ),
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **1** times:
`translate( 'Get Jetpack Security <em>Daily</em>' )` *ES Score: 6*

a8ci18n commented on client/my-sites/plans-v2/translated-copy.tsx:
```diff
@@ -0,0 +1,203 @@
+		case PLAN_JETPACK_SECURITY_DAILY_MONTHLY:
+			return {
+				displayName: translate( 'Security {{em}}Daily{{/em}}', { components: { em: <em /> } } ),
+				shortName: translate( 'Security Daily', {
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **19** times:
`translate( 'Security Daily', { context: 'Jetpack Security Daily Plan'} )` *ES Score: 15*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/fix/move-plansv2-translates-from-constants-take-2)

a8ci18n commented on client/my-sites/plans-v2/translated-copy.tsx:
```diff
@@ -0,0 +1,203 @@
+					'Enjoy the peace of mind of complete site protection. ' +
+						'Great for brochure sites, restaurants, blogs, and resume sites.'
+				),
+				buttonLabel: translate( 'Get Security {{em}}Daily{{/em}}', { components: { em: <em /> } } ),
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **20** times:
`translate( 'Security {{em}}Daily{{/em}}' )` *ES Score: 9*

a8ci18n commented on client/my-sites/plans-v2/translated-copy.tsx:
```diff
@@ -0,0 +1,203 @@
+		case PLAN_JETPACK_SECURITY_REALTIME_MONTHLY:
+			return {
+				displayName: translate( 'Security {{em}}Real-time{{/em}}', { components: { em: <em /> } } ),
+				shortName: translate( 'Security Real-time', {
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **19** times:
`translate( 'Security Real-time', { context: 'Jetpack Security Real-time Plan'} )` *ES Score: 13*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/fix/move-plansv2-translates-from-constants-take-2)

a8ci18n commented on client/my-sites/plans-v2/translated-copy.tsx:
```diff
@@ -0,0 +1,203 @@
+					'Additional security for sites with 24/7 activity. ' +
+						'Recommended for eCommerce stores, news organizations, and online forums.'
+				),
+				buttonLabel: translate( 'Get Security {{em}}Real-time{{/em}}', {
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **18** times:
`translate( 'Security {{em}}Real-time{{/em}}' )` *ES Score: 9*

rcanepa commented on client/my-sites/plans-v2/translated-copy.tsx:
```diff
@@ -0,0 +1,203 @@
+import {
+	PRODUCT_JETPACK_ANTI_SPAM,
+	PRODUCT_JETPACK_ANTI_SPAM_MONTHLY,
+	PRODUCT_JETPACK_BACKUP,
```
I think `PRODUCT_JETPACK_BACKUP` doesn't actually exist. What we use instead for options/generic cards is `OPTIONS_JETPACK_BACKUP` (and `OPTIONS_JETPACK_BACKUP_MONTHLY`). This fix an issue that appears if you change the term filter from annually to monthly.

robertf4 replied:
Good call!

rcanepa commented on client/my-sites/plans-v2/translated-copy.tsx:
```diff
@@ -0,0 +1,203 @@
+/**
+ * Internal dependencies
+ */
+import { OPTIONS_JETPACK_SECURITY, OPTIONS_JETPACK_SECURITY_MONTHLY } from './constants';
```
Here we need to add `OPTIONS_JETPACK_BACKUP` and `OPTIONS_JETPACK_BACKUP_MONTHLY`.

rcanepa commented on client/my-sites/plans-v2/translated-copy.tsx:
```diff
@@ -0,0 +1,203 @@
+				),
+				buttonLabel: translate( 'Get Anti-spam' ),
+			};
+		case PRODUCT_JETPACK_BACKUP:
```
If I'm not wrong, this one should be replaced with `OPTIONS_JETPACK_BACKUP` and `OPTIONS_JETPACK_BACKUP_MONTHLY`.

rcanepa commented on client/my-sites/plans-v2/translated-copy.tsx:
```diff
@@ -0,0 +1,203 @@
+	PRODUCT_JETPACK_CRM_MONTHLY,
+	PRODUCT_JETPACK_SCAN,
+	PRODUCT_JETPACK_SCAN_MONTHLY,
+} from 'lib/products-values/constants';
```
I noticed we also need to include Jetpack Search here.

robertf4 replied:
Good catch, included now.

rcanepa reviewed (changes requested):
I really like the idea behind this PR. I think it's good to remove the copy from the ProductSelector object.

I saw few issues:
- We need to include Jetpack Search in getSelectorProductCopy.
- The slug for the Jetpack Backup generic/option card is `OPTIONS_JETPACK_BACKUP` and not `PRODUCT_JETPACK_BACKUP`.

robertf4 commented:
Currently rebasing; PR has errors.

robertf4 committed:
Refactor copy out of utils

robertf4 committed:
Fix error

robertf4 commented:
Fixed errors, rebased again, more errors. Will return to this later.
